### PR TITLE
Add "slopspace" pool property

### PR DESCRIPTION
**Pull request by loli10K**
### Motivation and Context

This pool property can be set to override the reserved space that is normally determined using the system-wide kernel mode tunable "spa_slop_shift" to allow users to specify different per-pool settings
on systems sharing very large pools (e.g primary storage array) with smaller ones (e.g boot pool).

```
root@linux:~# truncate -s 512m /var/tmp/bpool
root@linux:~# truncate -s 128t /var/tmp/vmstorage
root@linux:~# zpool create bpool /var/tmp/bpool
root@linux:~# zpool create vmstorage /var/tmp/vmstorage
root@linux:~# df -h /vmstorage /bpool
Filesystem      Size  Used Avail Use% Mounted on
vmstorage       124T     0  124T   0% /vmstorage
bpool           352M     0  352M   0% /bpool
root@linux:~# 
root@linux:~# zpool set slopspace=10g vmstorage
root@linux:~# df -h /vmstorage /bpool
Filesystem      Size  Used Avail Use% Mounted on
vmstorage       128T     0  128T   0% /vmstorage
bpool           352M     0  352M   0% /bpool
root@linux:~# 
```

This was suggested during https://github.com/zfsonlinux/zfs/pull/7900 review.

### Description

 * man: add "slopspace" documentation
 * kernel: add "SLOP_SHIFT" definitions, change `spa_get_slop_space()` behaviour when "slopspace" is set
 * userland: pool property handling from the CLI
 * tests: add "slop_space" test group with 3 tests

### How Has This Been Tested?

Tests added to the ZTS

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ahrens** [comment]
What problem are you trying to solve?

I don't think this should be a supported, documented property.  It is exposing too much of the inner workings of ZFS.

**Comment by loli10K** [comment]
@ahrens  https://github.com/zfsonlinux/zfs/pull/7900#discussion_r218236087

**Comment by ahrens** [comment]
@loli10K Thank you for directing me to that comment.  However, it doesn't explain what problem this is solving.  What is wrong with the current behavior of the slop space (using 1/32nd of the pool)?

**Comment by loli10K** [comment]
This was specifically requested in the past for ZoL: https://github.com/zfsonlinux/zfs/issues/3014#issuecomment-227379381

> Would it be feasible to turn the setting into a pool-level property/tuneable?

>Asking since the spa_slop_shift being system wide this seems to be a bit unpractical as all systems I run have at least a small root pool and secondary big data pools, it would be helpful to be able to tune this per pool and not per system.

**Comment by richardelling** [comment]
IMHO, having a shift is a computationally efficient way to implement percentages. But historically, using percentages for reserved space in filesystems is short-sighted. 10 years later the technology will increase density by orders of magnitude and get much, much larger than the data stored. In other words, if I'm storing 1MB blocks, then I just need enough reserve to handle a burst of 1MB blocks. It has no direct relationship to the actual size of the pool.

In short, we need a cap on the reserve, not a percentage (implemented by shift) and not a per-pool percentage.

**Comment by DeHackEd** [comment]
Would it make sense for the slop space to be bounded? Like, no matter what the max slop space is 100 GB or something, and below that size you go with the module parameter?

I have a pool whose (raw) capacity is over 1/2 a petabyte. With `spa_slop_shift=5` that's ~17 terabytes of space reserved. That seems a bit much. Even with the old default of 6 that's still ~8 terabytes of reserved space.

You can argue that you're not supposed to fill a pool to the brink but 17 terabytes is a LOT.

**Comment by behlendorf** [comment]
Or to summarize there are few problems with the current `spa_slop_shift`.

* Reserving 1/32 of the pool is reasonable for small pools, but not for very large (1/2 PB) pools.
* It's global and not per-pool, making tuning it correctly impossible if your system has both a small and very large pool.
* It's not persistent.  Which makes it dodgy for us to ever change the default, as we saw previously.

Making this a pool property addresses most of these issues, but I agree the way its presented to end users is awkward.  I like @richardelling's idea of setting a reserved space cap as a pool property.  The idea of a filesystem needing some reserved space is something users are already be familiar with (ext, xfs, do this).  This could default to 1/32 initially so there no change by default.

**Comment by loli10K** [comment]
It seems the code in this PR is not the best way to handle these issues, probably an issue is more appropriate to discuss this.
